File: tests/device/hide/app-probe/app/src/main/java/dev/pathguard/hideprobe/JsonObservation.java
package dev.pathguard.hideprobe;

final class JsonObservation {
    private JsonObservation() {}

    static String render(
            String test,
            String surface,
            String path,
            long returnValue,
            int errno,
            boolean sideEffect,
            String status) {
        return "{\"schema\":1,\"kind\":\"observation\",\"test\":\""
                + escape(test)
                + "\",\"surface\":\""
                + escape(surface)
                + "\",\"path\":\""
                + escape(path)
                + "\",\"return_value\":"
                + returnValue
                + ",\"errno\":"
                + errno
                + ",\"side_effect\":"
                + sideEffect
                + ",\"status\":\""
                + escape(status)
                + "\"}";
    }

    static String escape(String value) {
        StringBuilder output = new StringBuilder(value.length());
        for (int index = 0; index < value.length(); ++index) {
            char ch = value.charAt(index);
            switch (ch) {
                case '"': output.append("\\\""); break;
                case '\\': output.append("\\\\"); break;
                case '\b': output.append("\\b"); break;
                case '\f': output.append("\\f"); break;
                case '\n': output.append("\\n"); break;
                case '\r': output.append("\\r"); break;
                case '\t': output.append("\\t"); break;
                default:
                    if (ch < 0x20) {
                        output.append(String.format("\\u%04x", (int) ch));
                    } else {
                        output.append(ch);
                    }
            }
        }
        return output.toString();
    }
}

File: tests/device/hide/app-probe/app/src/main/kotlin/dev/pathguard/hideprobe/JsonObservation.kt
package dev.pathguard.hideprobe

internal object JsonObservation {
    fun render(test: String, surface: String, path: String, returnValue: Long,
               errno: Int, sideEffect: Boolean, status: String): String =
        "{\"schema\":1,\"kind\":\"observation\",\"test\":\"${escape(test)}\"," +
            "\"surface\":\"${escape(surface)}\",\"path\":\"${escape(path)}\"," +
            "\"return_value\":$returnValue,\"errno\":$errno," +
            "\"side_effect\":$sideEffect,\"status\":\"${escape(status)}\"}"

    fun escape(value: String): String = buildString(value.length) {
        value.forEach { ch ->
            when (ch) {
                '"' -> append("\\\"")
                '\\' -> append("\\\\")
                '\b' -> append("\\b")
                '\u000c' -> append("\\f")
                '\n' -> append("\\n")
                '\r' -> append("\\r")
                '\t' -> append("\\t")
                else -> if (ch < ' ') append("\\u%04x".format(ch.code)) else append(ch)
            }
        }
    }

    fun renderStringArray(values: Array<String>) = values.joinToString(
        separator = ",", prefix = "[", postfix = "]",
    ) { "\"${escape(it)}\"" }
}
